### PR TITLE
WIP: Book Coverart and Grid Display

### PR DESCRIPTION
**Pull request by AbijeetP**
Towards #181

### CODE REVIEW

**Comment by nileshdeepak** [comment]
This pull request adds features for grid view and cover art [(#181)](https://github.com/BookStackApp/BookStack/issues/181).

The features that are provided are:

- Grid view for /books page.
- Choice of view type, grid or list, in the 'Edit Profile' page.
- Upload cover image at the time of creating or editing book.
- Displaying default cover if no image is uploaded.
- Toggle button to temporarily hide/show cover image in grid view.

Implementation for cover art required adding an ‘image’ column in the ‘books’ table for storing the id from the ‘images’ table. The image for grid view are stored with 'type' as 'cover'. The same convention has been followed for uploading the image as for app logo and user avatar.

The display type is stored in the ‘users’ table under ‘books_display’ and the default value is set to ‘grid’. A separate grid-list-item page has been added to implement the grid display. Users can select the desired display type, between grid and list, from the ‘Edit profile’ option any time.

Future enhancements:

- Grid view for other pages like home and index.
- Displaying cover image in list view.
- Editing image (cropping, resizing, etc)  before uploading.

Known issues:

- Even after the cover image is uploaded and selected, the selection resets when validation of other components fail.

Please review the code and do give your feedback.

File: tests/Entity/EntityTest.php [hunk 1]
```diff
@@ -145,16 +145,27 @@ public function bookCreation()
...
+            ->press('Select Image')
+            ->click('test-image.jpg')
+            ->press('Select Image')
             ->press('Save Book')
             // Check it redirects correctly
             ->seePageIs('/books/my-first-book')
             ->see($book->name)->see($book->description);
 
+         $book = factory(Book::class)->latest();
+        $this->assertDatabaseHas('images', [
```
Comment (AbijeetP) [inline tests/Entity/EntityTest.php]: not formatted.

File: app/Book.php [hunk 2]
```diff
@@ -18,6 +18,30 @@ public function getUrl($path = false)
         return baseUrl('/books/' . urlencode($this->slug));
     }
 
+    public function getBookCover()
```
Comment (bharadwajag) [inline app/Book.php]: Add proper comments to the function.

File: app/Book.php [hunk 3]
```diff
@@ -18,6 +18,30 @@ public function getUrl($path = false)
...
+    {
+        $default = baseUrl('/default.png');
+        $image = $this->image;
+        if ($image === 0 || $image === '0' || $image === null) 
+            return $default;
+        try {
+            $cover = $this->cover ? baseUrl($this->cover->getThumb(120, 192, false)) : $default;
```
Comment (bharadwajag) [inline app/Book.php]: - It's better to take width and height parameters instead of hard coding directly, so it will help us in feature to retrieve book cover in whatever size we want.

File: app/Book.php [hunk 4]
```diff
@@ -18,6 +18,30 @@ public function getUrl($path = false)
...
+        } catch (\Exception $err) {
+            $cover = $default;
+        }
+        return $cover;
+    }
+
+    public function getHeadingExcerpt($length = 35)
```
Comment (bharadwajag) [inline app/Book.php]: - Comments are missing.

File: app/Book.php [hunk 5]
```diff
@@ -18,6 +18,30 @@ public function getUrl($path = false)
...
+    {
+        $bookHeading = $this->name;
+        return strlen($bookHeading) > $length ? substr($bookHeading, 0, $length-3) . '...' : $bookHeading;
+    }
+    
+    public function cover()
+    {
+        return $this->belongsTo(Image::class, 'image');
```
Comment (bharadwajag) [inline app/Book.php]: - Added database column as **image**, but it should be **image_id**.

File: app/Http/Controllers/BookController.php [hunk 6]
```diff
@@ -36,11 +36,12 @@ public function __construct(EntityRepo $entityRepo, UserRepo $userRepo, ExportSe
      */
     public function index()
     {
-        $books = $this->entityRepo->getAllPaginated('book', 10);
+        $books = $this->entityRepo->getAllPaginated('book', 16);
```
Comment (bharadwajag) [inline app/Http/Controllers/BookController.php]: Changing of **10** to **16** books per page will effect in list grid also. 
I can understand your intention to made it as 16 ( 4 grid per row ), but it should not affect the list grid.

File: app/Http/Controllers/BookController.php [hunk 7]
```diff
@@ -36,11 +36,12 @@ public function __construct(EntityRepo $entityRepo, UserRepo $userRepo, ExportSe
...
         $recents = $this->signedIn ? $this->entityRepo->getRecentlyViewed('book', 4, 0) : false;
-        $popular = $this->entityRepo->getPopular('book', 4, 0);
+        $popular = $this->entityRepo->getPopular('book', 3, 0);
+        $books_display = $this->currentUser->books_display;
```
Comment (bharadwajag) [inline app/Http/Controllers/BookController.php]: `$books_display` should be `$booksDisplay`

File: database/migrations/2017_07_05_102650_add_cover_image_display.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('users', function (Blueprint $table) {
+            $table->string('books_display',10)->default('grid');
```
Comment (bharadwajag) [inline database/migrations/2017_07_05_102650_add_cover_image_display.php]: - `'books_display'` can we rename it to something better like '**books_view_type**'.
- Is data length 10 is okay for now?

File: database/migrations/2017_07_05_102650_add_cover_image_display.php [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        });
+
+        Schema::table('books', function (Blueprint $table) {
+            $table->integer('image');
```
Comment (bharadwajag) [inline database/migrations/2017_07_05_102650_add_cover_image_display.php]: - Column name should be image_id not image.

File: resources/assets/js/global.js [hunk 10]
```diff
@@ -99,6 +99,85 @@ jQuery.expr[":"].contains = $.expr.createPseudo(function (arg) {
     };
 });
 
+// Global jQuery Elements
```
Comment (bharadwajag) [inline resources/assets/js/global.js]: We can remove this part of the code, I guess it's related to notifications.

File: resources/assets/js/global.js [hunk 11]
```diff
@@ -99,6 +99,85 @@ jQuery.expr[":"].contains = $.expr.createPseudo(function (arg) {
...
+
+// Toggle thumbnail::hide image and reduce grid size
+$(document).ready(function(){
+   $('[data-action="expand-thumbnail"]').click(function(){
+     $('.gallery-item').toggleClass("collapse").find('img').slideToggle(50);
+   });
+});
+
+// Popup close
+$('.popup-close').click(function() {
```
Comment (bharadwajag) [inline resources/assets/js/global.js]: Not needed this part of code.

File: resources/assets/js/global.js [hunk 12]
```diff
@@ -99,6 +99,85 @@ jQuery.expr[":"].contains = $.expr.createPseudo(function (arg) {
...
+    }
+});
+
+// Common jQuery actions
+$('[data-action="expand-entity-list-details"]').click(function() {
+    $('.entity-list.compact').find('p').not('.empty-text').slideToggle(240);
+});
+
+// Toggle thumbnail::hide image and reduce grid size
+$(document).ready(function(){
```
Comment (bharadwajag) [inline resources/assets/js/global.js]: We don't need ` $(document).ready(function(){`

File: resources/views/components/image-picker.blade.php [hunk 13]
```diff
@@ -13,7 +13,7 @@
         <button class="text-button neg" data-action="remove-image" type="button">{{ trans('common.remove') }}</button>
     @endif
 
-    <input type="hidden" name="{{$name}}" id="{{$name}}" value="{{ isset($currentId) && ($currentId !== '' && $currentId !== false) ? $currentId : $currentImage}}">
+    <input type="hidden" name="{{$name}}" id="{{$name}}" value="{{ isset($currentId) && ($currentId !== 0 && $currentId !== false) ? $currentId : $currentImage}}">
```
Comment (bharadwajag) [inline resources/views/components/image-picker.blade.php]: Why this change` $currentId !== '' ` to `$currentId !== 0` ?

**Comment by bharadwajag** [comment]
@Abijeet @ssddanbrown 
This issue has been tracking under this pull request - https://github.com/BookStackApp/BookStack/pull/494
Here we can see the latest changes -  https://gfycat.com/UncomfortableSpanishHamadryad
